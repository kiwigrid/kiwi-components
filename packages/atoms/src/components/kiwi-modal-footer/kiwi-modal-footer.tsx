import {
  Component,
  Host,
  Event,
  EventEmitter,
  h,
  JSX,
  Prop,
  Fragment,
} from '@stencil/core';

/**
 * @slot kiwi-modal-footer-cancel - The content for the default cancel button.
 * @slot kiwi-modal-footer-confirm - The content for the default confirm button.
 */
@Component({
  tag: 'kiwi-modal-footer',
  shadow: false,
})
export class KiwiModalFooter {
  /**
   * The labels to be used for the default footer buttons, passed as a tuple of
   * strings with the first entry being the confirm label, and the second being
   * the cancel label.
   */
  @Prop()
  defaultLabels: [confirm: string, cancel: string] = ['Ok', 'Cancel'];

  /**
   * Whether to render the default footer buttons or a slot.
   *
   * This is a workaround since slot fallback content is not rendered correctly
   * with `shadow: false`
   */
  @Prop()
  useDefault = true;

  /** The event emitted when the user clicks the default cancel button. */
  @Event()
  cancel!: EventEmitter;

  /** The event emitted when the user clicks the default confirm button. */
  @Event()
  confirm!: EventEmitter;

  render(): JSX.Element {
    const [confirmLabel, cancelLabel] = this.defaultLabels;

    return (
      <Host slot="kiwi-modal-footer">
        <div class="modal-footer">
          {this.useDefault ? (
            <Fragment>
              <button class="btn btn-link" onClick={this.handleClose}>
                <slot name="kiwi-modal-footer-cancel">{cancelLabel}</slot>
              </button>
              <button class="btn btn-primary" onClick={this.handleConfirmation}>
                <slot name="kiwi-modal-footer-confirm">{confirmLabel}</slot>
              </button>
            </Fragment>
          ) : (
            <slot></slot>
          )}
        </div>
      </Host>
    );
  }

  private handleClose = (): void => {
    this.cancel.emit();
  };

  private handleConfirmation = (): void => {
    this.confirm.emit();
  };
}
