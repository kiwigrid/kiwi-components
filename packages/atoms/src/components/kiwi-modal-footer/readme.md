# kiwi-modal-footer

## Usage

### Default Labels

```typescriptreact
<kiwi-modal-footer
    onConfirm={() => /* confirm action*/}
    onCancel={() => /* cancel action*/}
/>
```

### Custom Labels

```typescriptreact
<kiwi-modal-footer
    defaultLabels={['Yes', 'Of course']}
    onConfirm={() => /* confirm action*/}
    onCancel={() => /* cancel action*/}
/>
```

### Custom Content For Default Buttons

```typescriptreact
<kiwi-modal-footer
    onConfirm={() => /* confirm action*/}
    onCancel={() => /* cancel action*/}
>
    <span slot="kiwi-modal-footer-confirm">…</span>
    <span slot="kiwi-modal-footer-cancel">…</span>
</kiwi-modal-footer>
```

### Custom Footer

```typescriptreact
<kiwi-modal-footer
    useDefault={false}
    onConfirm={() => /* confirm action*/}
    onCancel={() => /* cancel action*/}
>
    <button>Don't click me!</button>
</kiwi-modal-footer>
```

<!-- Auto Generated Below -->


## Properties

| Property        | Attribute     | Description                                                                                                                                                             | Type                                | Default            |
| --------------- | ------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ----------------------------------- | ------------------ |
| `defaultLabels` | --            | The labels to be used for the default footer buttons, passed as a tuple of strings with the first entry being the confirm label, and the second being the cancel label. | `[confirm: string, cancel: string]` | `['Ok', 'Cancel']` |
| `useDefault`    | `use-default` | Whether to render the default footer buttons or a slot.  This is a workaround since slot fallback content is not rendered correctly with `shadow: false`                | `boolean`                           | `true`             |


## Events

| Event     | Description                                                        | Type               |
| --------- | ------------------------------------------------------------------ | ------------------ |
| `cancel`  | The event emitted when the user clicks the default cancel button.  | `CustomEvent<any>` |
| `confirm` | The event emitted when the user clicks the default confirm button. | `CustomEvent<any>` |


## Slots

| Slot                          | Description                                 |
| ----------------------------- | ------------------------------------------- |
| `"kiwi-modal-footer-cancel"`  | The content for the default cancel button.  |
| `"kiwi-modal-footer-confirm"` | The content for the default confirm button. |


----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
