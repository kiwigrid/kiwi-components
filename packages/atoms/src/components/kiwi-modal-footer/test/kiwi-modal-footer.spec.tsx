import { h } from '@stencil/core';
import { newSpecPage } from '@stencil/core/testing';
import { KiwiModalFooter } from '../kiwi-modal-footer';

describe('kiwi-modal-footer', () => {
  it.each`
    name                                      | props                                  | children
    ${'default'}                              | ${{}}                                  | ${undefined}
    ${'custom labels'}                        | ${{ defaultLabels: ['ok', 'otmena'] }} | ${undefined}
    ${'default buttons with slotted content'} | ${{}}                                  | ${[<span slot="kiwi-modal-footer-confirm">Confirm</span>, <span slot="kiwi-modal-footer-cancel">Abort</span>]}
    ${'custom footer'}                        | ${{ useDefault: false }}               | ${(<button>A Button</button>)}
  `('renders $name', async ({ props, name, children }) => {
    const { root } = await newSpecPage({
      components: [KiwiModalFooter],
      template: () =>
        children ? (
          <kiwi-modal-footer {...props}>{children}</kiwi-modal-footer>
        ) : (
          <kiwi-modal-footer {...props} />
        ),
    });

    expect(root).toMatchSnapshot(name);
  });
});
