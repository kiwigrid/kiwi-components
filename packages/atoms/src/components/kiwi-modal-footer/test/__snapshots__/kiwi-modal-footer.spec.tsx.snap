// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`kiwi-modal-footer renders custom footer: custom footer 1`] = `
<kiwi-modal-footer slot="kiwi-modal-footer">
  <!---->
  <div class="modal-footer">
    <button>
      A Button
    </button>
  </div>
</kiwi-modal-footer>
`;

exports[`kiwi-modal-footer renders custom labels: custom labels 1`] = `
<kiwi-modal-footer slot="kiwi-modal-footer">
  <!---->
  <div class="modal-footer">
    <button class="btn btn-link">
      <slot-fb name="kiwi-modal-footer-cancel">
        otmena
      </slot-fb>
    </button>
    <button class="btn btn-primary">
      <slot-fb name="kiwi-modal-footer-confirm">
        ok
      </slot-fb>
    </button>
  </div>
</kiwi-modal-footer>
`;

exports[`kiwi-modal-footer renders default buttons with slotted content: default buttons with slotted content 1`] = `
<kiwi-modal-footer slot="kiwi-modal-footer">
  <!---->
  <div class="modal-footer">
    <button class="btn btn-link">
      <span slot="kiwi-modal-footer-cancel">
        Abort
      </span>
      <slot-fb hidden="" name="kiwi-modal-footer-cancel">
        Cancel
      </slot-fb>
    </button>
    <button class="btn btn-primary">
      <span slot="kiwi-modal-footer-confirm">
        Confirm
      </span>
      <slot-fb hidden="" name="kiwi-modal-footer-confirm">
        Ok
      </slot-fb>
    </button>
  </div>
</kiwi-modal-footer>
`;

exports[`kiwi-modal-footer renders default: default 1`] = `
<kiwi-modal-footer slot="kiwi-modal-footer">
  <!---->
  <div class="modal-footer">
    <button class="btn btn-link">
      <slot-fb name="kiwi-modal-footer-cancel">
        Cancel
      </slot-fb>
    </button>
    <button class="btn btn-primary">
      <slot-fb name="kiwi-modal-footer-confirm">
        Ok
      </slot-fb>
    </button>
  </div>
</kiwi-modal-footer>
`;
